subsys/nsettings/settings_fcb.c

```diff
@@ -0,0 +1,550 @@
+/* Copyright (c) 2019 Laczen
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ * Copyright (c) 2015 Runtime Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <string.h>
+
+#include "settings_fcb.h"
+#include "settings_priv.h"
+
+struct settings_fcb_read_fn_arg {
+	struct fcb_entry_ctx *entry_ctx;
+	off_t off;
+};
+
+static int settings_fcb_load(struct settings_store *cs);
+static int settings_fcb_save(struct settings_store *cs, const char *name,
+			     const char *value, size_t val_len);
+
+static struct settings_store_itf settings_fcb_itf = {
+	.csi_load = settings_fcb_load,
+	.csi_save = settings_fcb_save,
+};
+
+int settings_fcb_dst(struct settings_fcb *cf)
+{
+	cf->cf_store.cs_itf = &settings_fcb_itf;
+	settings_dst_register(&cf->cf_store);
+
+	return 0;
+}
+
+int settings_fcb_src(struct settings_fcb *cf)
+{
+	cf->cf_store.cs_itf = &settings_fcb_itf;
+	settings_src_register(&cf->cf_store);
+
+	return 0;
+}
+
+static int settings_fcb_read(struct fcb_entry_ctx *entry_ctx, off_t off,
+			     char *buf, size_t *len)
+{
+	if (off >= entry_ctx->loc.fe_data_len) {
+		return -EINVAL;
+	}
+
+	if ((off + *len) > entry_ctx->loc.fe_data_len) {
+		*len = entry_ctx->loc.fe_data_len - off;
+	}
+
+	return flash_area_read(entry_ctx->fap,
+			       FCB_ENTRY_FA_DATA_OFF(entry_ctx->loc) + off,
+			       buf, *len);
+}
+
+static ssize_t settings_fcb_read_fn(void *data, size_t len, void *read_fn_arg)
+{
+	int rc, len_read;
+	struct settings_fcb_read_fn_arg *rd_fn_arg;
+
+	len_read = len;
+	rd_fn_arg = (struct settings_fcb_read_fn_arg *)read_fn_arg;
+	rc = settings_fcb_read(rd_fn_arg->entry_ctx, rd_fn_arg->off,
+			       data, &len_read);
+	if (rc < 0) {
+		return rc;
+	}
+	return len_read;
+}
+
+static int settings_fcb_load(struct settings_store *cs)
+{
+	struct settings_fcb *cf = (struct settings_fcb *)cs;
+	struct fcb_entry_ctx loc1, loc2;
+	struct settings_fcb_read_fn_arg read_fn_arg;
+	struct settings_handler *ch;
+	char name1[SETTINGS_MAX_NAME_LEN + SETTINGS_EXTRA_LEN];
+	char name2[SETTINGS_MAX_NAME_LEN + SETTINGS_EXTRA_LEN];
+	char *name_argv[SETTINGS_MAX_DIR_DEPTH];
+	int rc, name_argc;
+	size_t len1_name, len2_name, len_data;
+	bool duplicate;
+
+	loc1.fap = cf->cf_fcb.fap;
+	loc1.loc.fe_sector = NULL;
+	loc1.loc.fe_elem_off = 0;
+
+	while (fcb_getnext(&cf->cf_fcb, &loc1.loc) == 0) {
+
+		len1_name = min(SETTINGS_MAX_NAME_LEN + SETTINGS_EXTRA_LEN + 1,
+				loc1.loc.fe_data_len);
+
+		rc = settings_fcb_read(&loc1, 0, name1, &len1_name);
+
+		if (rc) {
+			continue;
+		}
+
+		len1_name = strchr(name1, '=') - name1;
+
+		if (len1_name + 1 == loc1.loc.fe_data_len) {
+			/* Lack of a value so the record is a deletion-record
+			 * No sense to call h_set for a deleted record
+			 */
+			continue;
+		}
+
+		/* Find out if there is a later record with the same name */
+		loc2 = loc1;
+		duplicate = false;
+		while (fcb_getnext(&cf->cf_fcb, &loc2.loc) == 0) {
```

Comment:
Avoiding of manifestation of not recent values is great feature. But it increases settings load operation a lot - it becomes N<sup>2</sup> problem while was N.
My proposition is to make this optional (using Kconfig).
Very probably this affect FS back-end as well.

Great here is that above is not issue for NVS back-end.

Reply:
You are right that it needs to check more. The problem is N² on flash reads but the N load will be a lot bigger: the set needs to be processed each time (parse...), the user needs to check if it's receiving a deleted item, clear whatever it is doing on the set... I don't know what is worse.
Also it would mean that restoring factory settings would require a delete of the "normal" settings backend.

Reply:
After reconsideration I have decided to change it back to the original linear loading for fcb and file. It does however mean that the user needs to catch the deleted items.

Reply:
I would keep this as option. This will help to keep user code compatible betwean NVS and other back-ends.

Reply:
An option seems too complex. The user will need to catch the deleted items because it's not clear what backend will be selected (or how it is configured). This catch will be unnecessary for nvs, but this is not a problem.

Reply:
Isn't it just to make optional code form line114 to line137.

Reply:
Correct, it's not the code that is complex, it's the understanding of the configuration option. Making such an option is to difficult to understand, and even I cannot answer if it is advantageous or not to set the option, it depends on too many things: amount of entries in the backend, time to process a line, time to call set(). If I cannot answer this in 2 minutes how to decide to set the configuration option.

Reply:
Right, rather recommend to use other back-end ...so NVS. Agree.